File: PythonSolutions/Neetcode150/binarytree_problems/102_binary-tree-level-order-traversal.py
from typing import List
from collections import deque
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []

        res = []
        queue = deque([root])

        while queue:
            level = []
            n = len(queue)

            # Length of queue wont be re-calculated inside the for loop, allows to process each level
            for _ in range(n):
                curr = queue.popleft() # FIFO
                level.append(curr.val)

                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
            
            # Get the result of the specific level above
            res.append(level)

        return res

File: TypescriptSolutions/Neetcode150/binary-tree_problems/102_binary-tree-level-order-traversal.ts
import { Queue } from "@datastructures-js/queue";

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

const levelOrder = (root: TreeNode): number[][] => {
    if (!root) return [];

    const res: number[][] = [];
    const queue = new Queue([root]);

    while (!queue.isEmpty()) {
        const level: number[] = [];
        const n = queue.size();
        
        for (let i = 0; i < n; i++) {
            const curr = queue.dequeue()!;
            level.push(curr.val);

            if (curr.left) queue.enqueue(curr.left);
            if (curr.right) queue.enqueue(curr.right);
        }

        res.push(level);
    }

    return res;
};